packages/comlink/src/common.ts

```diff
@@ -0,0 +1,73 @@
+import {bufferCount, concatMap, defer, filter, fromEvent, map, pipe, take} from 'rxjs'
+import {fromEventObservable} from 'xstate'
+
+import type {ListenInput, ProtocolMessage} from './types'
+
+export const listenInputFromContext =
+  (
+    type: string | string[] = [],
+    options: {matches?: boolean; count?: number; responseType?: string} = {},
+  ) =>
+  <
+    T extends {
+      domain: string
+      connectTo: string
+      name: string
+      target: MessageEventSource | undefined
+    },
+  >({
+    context,
+  }: {
+    context: T
+  }): ListenInput => {
+    const {count, matches = true, responseType = 'message.received'} = options
+    return {
+      count,
+      domain: context.domain,
+      from: context.connectTo,
+      matches,
+      responseType,
+      target: context.target,
+      to: context.name,
+      type,
+    }
+  }
+
+export const listenFilter =
+  (input: ListenInput) =>
+  (event: MessageEvent<ProtocolMessage>): boolean => {
+    const {data} = event
+    const types = Array.isArray(input.type) ? input.type : [input.type]
+    return (
+      (input.matches ? types.includes(data.type) : !types.includes(data.type)) &&
+      data.domain === input.domain &&
+      data.from === input.from &&
+      data.to === input.to &&
+      (!input.target || event.source === input.target)
+    )
+  }
+
+export const eventToMessage =
+  <T>(type: T) =>
+  (event: MessageEvent<ProtocolMessage>): {type: T; message: MessageEvent<ProtocolMessage>} => ({
+    type,
+    message: event,
+  })
+
+export const messageEvents$ = defer(() =>
+  fromEvent<MessageEvent<ProtocolMessage>>(window, 'message'),
```

Comment:
rxjs + xstate = ❤

Reply:
Maybe we should setup a share replay here with deferred unsubscribe to handle cases where the <VisualEditing> component is unmounted and mounted again? This happens for example when moving between App Router and Pages Router in Next.js 🤔